sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNullAwareHashJoinExec.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.joins
+
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.optimizer.{BuildRight, BuildSide}
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.physical._
+import org.apache.spark.sql.execution.{ExplainUtils, SparkPlan}
+import org.apache.spark.sql.execution.metric.SQLMetrics
+import org.apache.spark.sql.internal.SQLConf
+
+private case class NotInSubqueryHashJoinParams(
+    buildSideRelation: HashedRelation,
+    buildSideIsNullExists: Boolean,
+    buildSideIsEmpty: Boolean,
+    streamedSideKey: Attribute,
+    streamedSideKeyIndex: Int)
+
+object NotInSubqueryConditionPattern {
```

Comment:
See comment above: Perhaps the pattern shouldn't be in the exec node.

In addition, please either create a case class for the tri-value return type or document them with a comment.

Reply:
I will move the case class and move the pattern code into 
org.apache.spark.sql.catalyst.planning.patterns.scala

same as ExtractEquiJoinKeys